### PR TITLE
[JENKINS-45126] fixed incremental build

### PR DESCRIPTION
**Pull request by daspilker**
@wolfs please take a look.

The manifest depends on the output of the compile tasks, so it can't be configured in the configuration phase. I created a "dynamic" manifest that computes it's values on demand by implementing Gradle's Manifest interface. I don't see a better solution without accessing internal classes. Suggestions for a nicer solution are welcome.

### CODE REVIEW

File: src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiManifest.groovy [hunk 1]
```diff
@@ -68,35 +78,89 @@ class JpiManifest extends Manifest {
...
         YesNoMaybe supportDynamicLoading = isSupportDynamicLoading(classDir)
         if (supportDynamicLoading != YesNoMaybe.MAYBE) {
-            mainAttributes.putValue('Support-Dynamic-Loading', (supportDynamicLoading == YesNoMaybe.YES).toString())
+            mainAttributes['Support-Dynamic-Loading'] = (supportDynamicLoading == YesNoMaybe.YES).toString()
         }
 
         // remove empty values
         mainAttributes.entrySet().removeAll { it.value == null || it.value.toString().empty }
+
+        mainAttributes
```
Comment (wolfs) [inline src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiManifest.groovy]: I prefer using an explicit `return` keyword instead of writing a variable on the last line.

File: src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiManifest.groovy [hunk 2]
```diff
@@ -68,35 +78,89 @@ class JpiManifest extends Manifest {
...
+    }
+
+    @Override
+    @SuppressWarnings('Instanceof')
+    Manifest writeTo(Object o) {
+        if (!(o instanceof OutputStream)) {
+            throw new IllegalArgumentException()
+        }
+        toJavaManifest().write(o)
+        this
```
Comment (wolfs) [inline src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiManifest.groovy]: I prefer using an explicit `return` keyword instead of writing a variable on the last line.

File: src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiManifest.groovy [hunk 3]
```diff
@@ -68,35 +78,89 @@ class JpiManifest extends Manifest {
...
+    Manifest from(Object o, Closure<?> closure) {
+        throw new UnsupportedOperationException()
+    }
+
+    java.util.jar.Manifest toJavaManifest() {
+        java.util.jar.Manifest manifest = new java.util.jar.Manifest()
+        attributes.each {
+            manifest.mainAttributes.putValue(it.key, it.value.toString())
+        }
+        manifest
```
Comment (wolfs) [inline src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiManifest.groovy]: I prefer using an explicit `return` keyword instead of writing a variable on the last line.

File: src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiManifest.groovy [hunk 4]
```diff
@@ -109,13 +173,13 @@ class JpiManifest extends Manifest {
     }
 
     private static listUpDependencies(Configuration c, boolean optional, StringBuilder buf) {
-        for (ResolvedArtifact a : c.resolvedConfiguration.resolvedArtifacts) {
+        for (Dependency d : c.dependencies) {
```
Comment (wolfs) [inline src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiManifest.groovy]: Is this an unrelated change?

File: src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiManifest.groovy [hunk 5]
```diff
@@ -137,7 +201,6 @@ class JpiManifest extends Manifest {
         YesNoMaybe.YES
     }
 
-    static Map<String, ?> attributesToMap(Attributes attributes) {
-        attributes.collectEntries { k, v -> [k.toString(), v] } as Map<String, ?>
+    static class JpiAttributes extends HashMap<String, Object> implements Attributes {
```
Comment (wolfs) [inline src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiManifest.groovy]: I think it would be better to use a `LinkedHashMap` to have a repeatable order of elements.

**Comment by wolfs** [comment]
In the Gradle build itself we are also using a configure task to set attributes of the manifest which are only available at execution time: https://github.com/gradle/gradle/blob/6bbb9d6c6b0e9efc7f1dbcb51b9086741b488dc0/subprojects/launcher/launcher.gradle#L40-L44

File: src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiManifest.groovy (hunk 4 above)
Comment (daspilker) [inline src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiManifest.groovy]: Yes that's an unrelated change aka merge error.

**Comment by daspilker** [comment]
@wolfs thanks for the review. I thought about changing the task configuration at execution time but considered it bad style since I never saw that. But now it has your official approval, so I will take that approach. 😃

**Comment by daspilker** [comment]
Closing this in favor of #89.
